yapapi/script/__init__.py

```diff
@@ -0,0 +1,201 @@
+from datetime import timedelta
+import itertools
+from typing import Any, Awaitable, Callable, Dict, Iterator, Optional, List, TYPE_CHECKING
+
+from yapapi.events import CommandExecuted
+from yapapi.script.capture import CaptureContext
+from yapapi.script.command import (
+    BatchCommand,
+    Command,
+    Deploy,
+    DownloadBytes,
+    DownloadFile,
+    DownloadJson,
+    Run,
+    SendBytes,
+    SendFile,
+    SendJson,
+    Start,
+    Terminate,
+)
+from yapapi.storage import DOWNLOAD_BYTES_LIMIT_DEFAULT
+
+if TYPE_CHECKING:
+    from yapapi.ctx import WorkContext
+
+script_ids: Iterator[int] = itertools.count(1)
+"""An iterator providing incremental integer IDs to scripts."""
+
+
+class Script:
+    """Represents a series of commands to be executed on a provider node.
+
+    New commands are added to the script either through its `add` method or by calling one of the
+    convenience methods provided (for example: `run` or `upload_json`).
+    Adding a new command *does not* result in it being immediately executed. Once ready, a `Script`
+    instance is meant to be yielded from a worker function (work generator pattern).
+    Commands will be run in the order in which they were added to the script.
+
+    If the `WorkContext` instance this `Script` uses has the field `_implicit_init` set to `True`,
+    the first yielded script is going to be prepended with `Deploy` and
+    `Start` commands.
+    """
+
+    timeout: Optional[timedelta]
+    """Time after which this script's execution should be forcefully interrupted.
+
+    The default value is `None` which means there's no timeout set.
+    """
+
+    wait_for_results: bool
+    """Whether this script's execution should block until its results are available.
+
+    The default value is `True`.
+    """
+
+    def __init__(
+        self,
+        context: "WorkContext",
+        timeout: Optional[timedelta] = None,
+        wait_for_results: bool = True,
+    ):
+        self.timeout = timeout
+        self.wait_for_results = wait_for_results
```

Comment:
Hmm, I've never seen this way of documenting object fields.
I think the standard way is:

```suggestion
    def __init__(
        self,
        context: "WorkContext",
        timeout: Optional[timedelta] = None,
        wait_for_results: bool = True,
    ):
        self.timeout = timeout
        """Time after which this script's execution should be forcefully interrupted.

        The default value is `None` which means there's no timeout set.
        """

        self.wait_for_results = wait_for_results
        """Whether this script's execution should block until its results are available.

        The default value is `True`.
        """

```

Reply:
I find this more readable than having docstrings inside `__init__`, we've used this convention across `goth` (for example: https://github.com/golemfactory/goth/blob/2b40c505fe79ef3238880114e4dddfab2cfcc6aa/goth/runner/container/__init__.py#L50). I'd keep the docstrings this way as long as they work as intended with `sphinx`.

Reply:
OK, that makes sense.
I don't know if the work with `sphinx`, but if this is an accepted `python` convention I think they will.